Blob info caching + mount/reuse

mtrmac opened the pull request:
This adds types.BlobInfoCache recording the encountered compressed/uncompressed digest relationships and `docker://` blob locations, and using them to:
- Mount blobs on `docker://` destinations instead of copying the data again
- Reuse local layers on `containers-storage:` destinations instead of pulling them
- Avoid blob decompression when converting from schema1 just to figure out the DiffID values

The blob info cache is stored in `/var/lib/containers/blob-info-cache-v1.boltdb` for EUID=0 users, and `$XDG_DATA_HOME/containers/blob-info-cache-v1.boltdb` for unprivileged users.

See individual commit messages for more details.

WIP items:
- <s>When `canSubstitute`, `dockerImageDestination` makes a mount request even within the same repo, which is unnecessary.</s>
- <s>Documentation of the `BlobInfoCache` interface and possibly other code</s>
- Tests for corner cases of the BoltDB code
- Logging on failures in the BoltDB code

rhatdan commented:
@vrothberg @runcom PTAL

rhatdan commented:
@nalind PTAL

mtrmac commented:
Tests are failing here due to the API change, updated Skopeo is running the tests in https://github.com/containers/skopeo/pull/570 .

mtrmac commented:
> The blob info cache is stored in `/var/lib/containers/blob-info-cache-v1.boltdb` for EUID=0 users, and `$XDG_DATA_HOME/containers/blob-info-cache-v1.boltdb` for unprivileged users.

This might not be the right thing to do - apparently rootless buildah runs in a user namespace, which causes it to see EUID==0, and try to write to `/var/lib/containers`.  That does not _break_ builds because any error causes the code to fall back to memory-only caches, but it makes the optimizations unavailable.

rhatdan commented on pkg/blobinfocache/boltdb.go:
```diff
@@ -0,0 +1,310 @@
+		}
+		return nil
+	}); err != nil { // Including os.IsNotExist(err)
+		return []types.BICReplacementCandidate{} // FIXME? Log?
```
Shouldn't we log everything except IsNotExist?

rhatdan commented on copy/copy.go:
```diff
@@ -502,28 +502,19 @@ type diffIDResult struct {
+			return types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+		}
+		if reused {
+			ic.c.Printf("Skipping fetch of repeat blob %s\n", srcInfo.Digest)
```
Can you reword this?

Skipping blob %s fetch, already exists?
Is this something we actually need to see, or would it be better to just see what is pulled?

mtrmac replied:
(“fetch” in this message includes pushes.)

This can definitely be improved, at least `s/blob/layer/`, and maybe we don’t need the digests in the output at all. That’s all fairly independent from the functionality of this PR.

mtrmac replied:
> (“fetch” in this message includes pushes.)

i.e. all of pull/push/copy go through this code path and may skip transferring a blob between transports here.

rhatdan replied:
Does this just become noise on the screen, would the user care?  Seems to be more debugging information.

runcom replied:
what about `Blob transfer skip` - more generic than fetch and more valid in the context of pushes

rhatdan commented:
@vrothberg @runcom @nalind @umohnani8 PTAL

rhatdan commented:
@mtrmac Is this still WIP?  Can we get it vendored into skopeo so we can get tests to pass?

mtrmac commented:
> @mtrmac Is this still WIP?

Yes, at least for documentation.

> Can we get it vendored into skopeo so we can get tests to pass?

Already exists in 
https://github.com/containers/skopeo/pull/570 .

mtrmac commented on pkg/blobinfocache/boltdb.go:
```diff
@@ -0,0 +1,310 @@
+			return err
+		}
+		return nil
+	}) // FIXME? Log error?
```
Yes, probably. That’s one of the WIP parts.

mtrmac replied:
(The a bit tricky aspect to this is that failures using the cache are _never_ fatal (which might not be a good idea, to be fair), so a failure could result in dozens of identical or similar error messages per copy; we probably want to throttle them somehow, or maybe even mark the cache as broken after a few failures.)

rhatdan commented:
I'm willing to take it without Documentation, so we can move the testing along. Then open a separate PR for Docs

Why is the tests failing?

mtrmac commented:
Updated to avoid a no-op mount within a single repo, and added documentation to the interface/implementations; I can live with this being merged.

Tests fail here due to the API change, should eventually pass in https://github.com/containers/skopeo/pull/570 (and after merging this, a variant of https://github.com/containers/skopeo/pull/570 that vendors from `master` of this repo will need to be quickly merged).

rhatdan commented:
LGTM
@giuseppe @vrothberg @nalind @runcom @umohnani8 PTAL

runcom commented:
Reviewing this in a bit

runcom commented:
I had just a comment which we can live with for now/beginning of January.
This PR also passes our demo and code is pretty solid as well.

I'm gonna merge this, if you have anything against, please fill out issues/PRs

runcom commented:
Tests in skopeo pass as well https://github.com/containers/skopeo/pull/570
